src/pg_query_summary.c

```diff
@@ -0,0 +1,765 @@
+		summary->tables = state.tables;
+
+	if (state.aliases != NULL)
+		summary->aliases = state.aliases;
+
+	if (state.cte_names != NULL)
+		summary->cte_names = state.cte_names;
+
+	if (state.functions != NULL)
+		summary->functions = state.functions;
+
+	if (state.filter_columns != NULL)
+		summary->filter_columns = state.filter_columns;
+}
+
+// Given a walked parse tree and summary, store the truncated version in `summary`.
+//
+// Returns NULL on success.
+static PgQueryError* summary_truncate(Summary* summary, Node* tree)
+{
+	return make_pg_query_error("summary_truncate() is not implemented yet");
+
+	// TODO: implement summary_truncate()
+
+	// No error to return.
+	return NULL;
+}
+
+// Given the summarized query, convert it to protobuf, and store it in `result`.
+static void summary_to_protobuf(PgQuerySummaryParseResult* result, Summary* summary)
+{
+	PgQueryProtobuf protobuf;
+	PgQuery__SummaryResult sr = PG_QUERY__SUMMARY_RESULT__INIT;
+
+	sr.n_tables = list_length(summary->tables);
+	sr.tables = palloc(sizeof(PgQuery__SummaryResult__Table*) * sr.n_tables);
+
+	ListCell *lc = NULL;
+	foreach(lc, summary->tables) {
+		size_t i = foreach_current_index(lc);
+		SummaryTable* table = lfirst(lc);
+		sr.tables[i] = palloc(sizeof(PgQuery__SummaryResult__Table));
+		pg_query__summary_result__table__init(sr.tables[i]);
+		sr.tables[i]->name = table->name;
+		sr.tables[i]->schema_name = table->schema_name;
+		sr.tables[i]->table_name = table->table_name;
+		sr.tables[i]->context = (PgQuery__SummaryResult__Context)table->context;
+	}
+
+	sr.n_aliases = list_length(summary->aliases);
+	sr.aliases = palloc(sizeof(PgQuery__SummaryResult__AliasesEntry*) * sr.n_aliases);
+	foreach(lc, summary->aliases) {
+		size_t i = foreach_current_index(lc);
+		SummaryAlias* alias = lfirst(lc);
+
+		sr.aliases[i] = palloc(sizeof(PgQuery__SummaryResult__AliasesEntry));
+		pg_query__summary_result__aliases_entry__init(sr.aliases[i]);
+		sr.aliases[i]->key = alias->key;
+		sr.aliases[i]->value = alias->value;
+	}
+
+	sr.n_cte_names = list_length(summary->cte_names);
+	sr.cte_names = palloc(sizeof(char*) * sr.n_cte_names);
+	foreach(lc, summary->cte_names) {
+		size_t i = foreach_current_index(lc);
+		char *ctename = lfirst(lc);
+		sr.cte_names[i] = ctename;
+	}
+
+	sr.n_functions = list_length(summary->functions);
+	sr.functions = palloc(sizeof(PgQuery__SummaryResult__Function*) * sr.n_functions);
+	foreach(lc, summary->functions) {
+		size_t i = foreach_current_index(lc);
+		SummaryFunction* fn = lfirst(lc);
+
+		sr.functions[i] = palloc(sizeof(PgQuery__SummaryResult__Function));
+		pg_query__summary_result__function__init(sr.functions[i]);
+		sr.functions[i]->name = fn->name;
+		sr.functions[i]->function_name = fn->function_name;
+		sr.functions[i]->schema_name = fn->schema_name;
+		// FIXME: Implicit cast from enum ContextType to enum _PgQuery__SummaryResult__Context.
+		// This can't actually be guaranteed as correct -- probably needs actual logic, sadly.
```

Comment:
As discussed, this would be good to take care of now, so we don't forget about it. Since this is not performance critical having explicit mapping logic with a switch statement seems fine.